File: backend/src/routes/events.rs
//! Server-sent events: one long-lived connection per client, replacing polling.
//!
//! SSE rather than WebSockets, for three reasons specific to this deployment.
//! The traffic is one-directional — the server has nothing to receive that an
//! ordinary request does not already carry. It survives the existing edge
//! unchanged, with no protocol upgrade for Cloudflare or nginx to negotiate.
//! And reconnection is part of the format, so a dropped connection is the
//! client's problem to retry rather than a state machine to write.
//!
//! # Authentication
//!
//! The browser `EventSource` API cannot set an `Authorization` header, which is
//! why so many SSE endpoints end up taking a token in the query string. That is
//! not done here: query strings reach access logs, `Referer` headers and browser
//! history, and this repository already went to some length to keep the calendar
//! feed token out of exactly those places. Clients stream this with `fetch`
//! instead, which carries the header normally.

use std::time::Duration;

use actix_web::{web, HttpRequest, HttpResponse};
use futures_util::stream;
use tokio::sync::broadcast::error::RecvError;
use uuid::Uuid;

use crate::config::Config;
use crate::errors::AppError;
use crate::middleware::auth::require_auth;
use crate::services::events::{self, UserEvent};
use crate::services::revocation;
use crate::utils::jwt;

/// How many simultaneous streams one account may hold.
///
/// A stream costs a task and a socket for as long as it is open, so without a
/// bound a single account could pin them by opening tabs.
///
/// This used to be five, to match the active-session cap. That was the wrong
/// comparison — a session is a signed-in browser, and one browser holds as many
/// streams as it has tabs — and it interacted badly with how a dead stream is
/// noticed. The count comes from live `broadcast::Receiver`s, and the receiver
/// for an abandoned stream survives until this process next tries to write to
/// it, which is at most `KEEPALIVE_INTERVAL` away. So a reload leaves its old
/// stream counted for a while, and refreshing a few times in quick succession
/// refused the *live* connection because of ones the browser had already
/// dropped: 429, and no live updates until the ghosts timed out.
///
/// The cap therefore has to clear the number of reconnects a person can make
/// inside one detection window, with room to spare. Ten against a ten-second
/// window is not a threshold ordinary use reaches; it is still a hard bound on
/// sockets and tasks, which is the only thing this was ever for.
const MAX_CONNECTIONS_PER_USER: usize = 10;

/// Idle gap after which a comment frame is emitted.
///
/// Proxies close connections that go quiet, and a client cannot tell a healthy
/// idle stream from a dead one. Comment frames are ignored by the SSE parser,
/// so this is invisible to the application and only keeps the pipe warm.
///
/// It is also, unavoidably, how long it takes to notice a client that left.
/// Server-sent events have no liveness probe of their own: this process learns
/// the far end is gone when a write fails, and this is when it writes. That
/// makes the interval the detection window the cap above has to accommodate,
/// which is why it is ten seconds rather than twenty-five. The extra traffic is
/// a twelve-byte comment frame every ten seconds per open stream.
const KEEPALIVE_INTERVAL: Duration = Duration::from_secs(10);

/// The cap, for the tests that assert what it has to clear.
///
/// Exposed rather than duplicated: a test carrying its own copy of the number
/// passes while the code it guards says something else.
pub fn max_connections_per_user() -> usize {
    MAX_CONNECTIONS_PER_USER
}

/// The detection window, for the same reason.
pub fn keepalive_interval() -> Duration {
    KEEPALIVE_INTERVAL
}

pub fn configure(cfg: &mut web::ServiceConfig) {
    cfg.route("/events", web::get().to(stream_events));
}

/// A guard that frees the user's channel when the connection ends, however it
/// ends — returned normally, client disconnected, or the task was dropped.
/// Doing this in `Drop` rather than after the loop is what makes it hold for
/// the disconnect case, which is the common one.
struct ConnectionGuard {
    user_id: Uuid,
    /// The session and issue time this stream was opened with, so revocation
    /// can be re-checked while it is open.
    session_id: Uuid,
    issued_at: i64,
    issued_at_ms: Option<i64>,
}

impl Drop for ConnectionGuard {
    fn drop(&mut self) {
        events::release(self.user_id);
        // Paired with the increment where the guard is built, so every way a
        // stream can end — client gone, session revoked, shutdown — decrements
        // exactly once.
        crate::metrics::record_event_stream_closed();
    }
}

fn frame(event: UserEvent) -> String {
    // `event:` lets a client register one handler per kind. The data payload is
    // still sent so a generic listener has something well-formed to parse, and
    // because a frame without a data line is not delivered at all.
    format!(
        "event: {}\ndata: {{\"kind\":\"{}\"}}\n\n",
        event.name(),
        event.name()
    )
}

/// Re-read the validated claims for this request.
///
/// `require_auth` returns only the user id, which is all an ordinary handler
/// needs. A stream also needs the session id and issue time, because it has to
/// answer the revocation question again later, and the token is not available
/// once the response has started.
fn session_claims(req: &HttpRequest) -> Result<jwt::Claims, AppError> {
    let config = req
        .app_data::<web::Data<Config>>()
        .ok_or_else(|| AppError::Unauthorized("Server configuration error".to_string()))?;
    let token = req
        .headers()
        .get(actix_web::http::header::AUTHORIZATION)
        .and_then(|value| value.to_str().ok())
        .and_then(|value| value.strip_prefix("Bearer "))
        .ok_or_else(|| AppError::Unauthorized("Missing Authorization header".to_string()))?;
    jwt::validate_token(token, &config.jwt_secret)
}

async fn stream_events(req: HttpRequest) -> Result<HttpResponse, AppError> {
    let user_id = require_auth(&req).await?;

    if events::connection_count(user_id) >= MAX_CONNECTIONS_PER_USER {
        return Err(AppError::TooManyRequests(
            "Too many open event streams for this account".to_string(),
        ));
    }

    // Ordinary requests are authorized once because they are over in
    // milliseconds. This one lives for as long as a tab stays open, so a single
    // check at the handshake would make it the one credential that outlives
    // "sign out everywhere" — the exact hole access-token revocation closed.
    // Keep the claims so the stream can re-check itself as it runs.
    let claims = session_claims(&req)?;

    let receiver = events::subscribe(user_id);
    let guard = ConnectionGuard {
        user_id,
        session_id: claims.sid,
        issued_at: claims.iat,
        issued_at_ms: claims.iat_ms,
    };
    // After the guard exists, never before: anything that could fail between
    // the two would leave the gauge counting a stream nobody holds.
    crate::metrics::record_event_stream_opened();

    let body = stream::unfold((receiver, guard), |(mut receiver, guard)| async move {
        let next = tokio::time::timeout(KEEPALIVE_INTERVAL, receiver.recv()).await;

        // Checked on every wake-up, including keepalives, so a revoked session
        // is dropped within one keepalive interval at the latest rather than
        // whenever the client happens to reconnect.
        if revocation::is_revoked(
            guard.session_id,
            guard.user_id,
            guard.issued_at,
            guard.issued_at_ms,
        ) {
            return None;
        }

        let chunk = match next {
            Ok(Ok(event)) => frame(event),
            // The connection fell behind its buffer. Say so rather than
            // resuming silently: the client has missed an unknown number
            // of hints, and the honest recovery is a full refetch, which
            // is what it does when it sees this.
            Ok(Err(RecvError::Lagged(missed))) => {
                log::debug!(
                    "event stream lagged user_id={} missed={missed}",
                    guard.user_id
                );
                ": lagged\n\n".to_string()
            }
            // Every sender is gone, which only happens at shutdown.
            Ok(Err(RecvError::Closed)) => return None,
            // Idle. A comment frame keeps proxies from reaping the
            // connection and lets the client tell alive from stalled.
            Err(_elapsed) => ": keepalive\n\n".to_string(),
        };

        Some((
            Ok::<_, actix_web::Error>(web::Bytes::from(chunk)),
            (receiver, guard),
        ))
    });

    Ok(HttpResponse::Ok()
        .content_type("text/event-stream")
        // `no-store` for the same reason every authenticated response carries
        // it; `no-transform` stops a CDN rewriting the body mid-stream.
        .insert_header(("Cache-Control", "no-store, no-transform"))
        // nginx buffers proxied responses by default, which would hold every
        // frame until the buffer filled — turning a push into a very slow poll.
        .insert_header(("X-Accel-Buffering", "no"))
        .streaming(body))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[actix_web::test]
    async fn an_open_stream_is_dropped_once_its_session_is_revoked() {
        // The property that makes a long-lived connection safe. Without it the
        // stream would be the one credential that survives "sign out
        // everywhere", because it is authorized once and then runs for hours.
        let user_id = Uuid::new_v4();
        let session_id = Uuid::new_v4();
        let issued_at = chrono::Utc::now().timestamp();

        assert!(!revocation::is_revoked(
            session_id, user_id, issued_at, None
        ));
        revocation::revoke_session_in_memory(session_id);
        assert!(
            revocation::is_revoked(session_id, user_id, issued_at, None),
            "the stream's own liveness check must see the revocation"
        );
    }

    #[test]
    fn a_frame_is_well_formed_and_names_its_event() {
        let rendered = frame(UserEvent::MessagesChanged);
        assert_eq!(
            rendered,
            "event: messages\ndata: {\"kind\":\"messages\"}\n\n"
        );
        // The blank line is what terminates an SSE frame; without it the client
        // buffers indefinitely and nothing is ever delivered.
        assert!(rendered.ends_with("\n\n"));
    }

    #[test]
    fn frames_carry_no_content() {
        // The whole authorization argument rests on this: the payload names a
        // kind and nothing else. A future field carrying, say, a message body
        // would need every read-path rule re-checked here.
        for event in [UserEvent::MessagesChanged, UserEvent::NotificationsChanged] {
            let rendered = frame(event);
            assert!(rendered.contains(event.name()));
            assert_eq!(rendered.matches("data:").count(), 1);
            assert!(!rendered.contains("body"));
            assert!(!rendered.contains("username"));
        }
    }
}

File: frontend/src/lib/events.ts
/**
 * Client for the server-sent event stream.
 *
 * Built on `fetch` rather than `EventSource`, because `EventSource` cannot send
 * an `Authorization` header. The usual workaround is a token in the query
 * string, which is exactly what this codebase avoids elsewhere: query strings
 * reach access logs, `Referer` headers and browser history.
 *
 * The events carry no content — only which area changed — so every handler here
 * does the same thing: invalidate and refetch through the normal authenticated
 * endpoints. That keeps one authorization path and makes a dropped event cost a
 * delayed refresh rather than a missing message.
 */

import { refreshAccessToken } from '@/lib/api';

const RECONNECT_BASE_MS = 1_000;
const RECONNECT_MAX_MS = 30_000;

export type ServerEventKind = 'messages' | 'notifications';

export interface EventStreamHandlers {
  onEvent: (kind: ServerEventKind) => void;
  /** Called whenever a stream starts, including after a reconnect. The client
   *  cannot know what it missed while disconnected, so this is the signal to
   *  refetch everything the stream would otherwise have told it about. */
  onResync: () => void;
}

/** Parse one SSE frame. Comment lines (`:` prefixed) are keepalives and carry
 *  no event, so they yield null. */
function parseFrame(frame: string): ServerEventKind | null {
  let name: string | null = null;
  for (const line of frame.split('\n')) {
    if (line.startsWith('event:')) name = line.slice('event:'.length).trim();
  }
  return name === 'messages' || name === 'notifications' ? name : null;
}

/**
 * Hold a stream open, reconnecting with backoff until `stop` is called.
 *
 * Returns a stop function. Reconnection is unconditional by design: the server
 * closes the stream when a session is revoked, and the retry that follows will
 * fail authentication and surface through the normal 401 path rather than
 * looping silently.
 */
export function connectEventStream(
  apiUrl: string,
  getToken: () => string | null,
  handlers: EventStreamHandlers,
): () => void {
  let stopped = false;
  let controller: AbortController | null = null;
  let attempt = 0;
  let timer: ReturnType<typeof setTimeout> | null = null;

  const run = async (): Promise<void> => {
    const token = getToken();
    if (!token) return schedule();

    controller = new AbortController();
    try {
      const response = await fetch(`${apiUrl}/api/events`, {
        headers: { Authorization: `Bearer ${token}` },
        signal: controller.signal,
        credentials: 'include',
      });
      if (response.status === 401) {
        // The access token expired while the stream was open, or between a
        // reconnect and this attempt. Retrying with the same token cannot
        // succeed: this request never touches the axios interceptor that
        // refreshes, because a streaming body is not something axios returns.
        //
        // Without this the stream sat in a 401 loop every thirty seconds until
        // some unrelated query happened to refresh the session — which is
        // exactly what a backend restart produced, in every open tab at once.
        try {
          await refreshAccessToken();
        } catch {
          // The session is genuinely gone. Back off rather than hammer.
          return schedule();
        }
        // A refreshed session is not a failed attempt; reconnect promptly
        // rather than waiting out a backoff earned by something else.
        attempt = 0;
        return schedule();
      }
      if (!response.ok || !response.body) return schedule();

      // Connected. Anything that happened while disconnected is unknown, so
      // resync before processing the first frame.
      attempt = 0;
      handlers.onResync();

      const reader = response.body.getReader();
      const decoder = new TextDecoder();
      let buffer = '';

      for (;;) {
        const { done, value } = await reader.read();
        if (done || stopped) break;
        buffer += decoder.decode(value, { stream: true });

        // Frames are separated by a blank line; anything after the last one is
        // a partial frame and stays buffered.
        const frames = buffer.split('\n\n');
        buffer = frames.pop() ?? '';
        for (const frame of frames) {
          // A `: lagged` comment means the server dropped signals for this
          // connection. It cannot say which, so treat it like a reconnect.
          if (frame.startsWith(': lagged')) {
            handlers.onResync();
            continue;
          }
          const kind = parseFrame(frame);
          if (kind) handlers.onEvent(kind);
        }
      }
    } catch {
      // Network error, navigation, or an aborted request. Backoff handles it.
    }
    return schedule();
  };

  const schedule = (): void => {
    if (stopped) return;
    // Exponential with a ceiling, so a backend restart does not turn every open
    // tab into a tight retry loop against a service that is still starting.
    const delay = Math.min(RECONNECT_BASE_MS * 2 ** attempt, RECONNECT_MAX_MS);
    attempt += 1;
    timer = setTimeout(() => void run(), delay);
  };

  void run();

  return () => {
    stopped = true;
    if (timer) clearTimeout(timer);
    controller?.abort();
  };
}
